### ISSUE TITLE
Remove app.use()

### ISSUE THREAD

**Issue by mcollina** [fastify/fastify#3502]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the feature has not already been requested

### 🚀 Feature Proposal

I think we should be removing `app.use()` in the `next` release, i.e. https://github.com/fastify/fastify/blob/a17482959d4a4a9fab31926d0443ecd13d078c99/fastify.js#L327-L330.

### Motivation

We should not be encouraging people to use middlewares. Both `middie` and `fastify-express` will still be there and maintained.

**Comment by jsumners** [comment]
Please do this. I wouldn't be upset if we also deprecate `middie` and `fastify-express` with v4.

**Comment by mcollina** [comment]
I think they are useful plugins, we should keep having them but reduce their visibility.

**Comment by genzyy** [comment]
Can I take this? How should I go about this like removing `use` will do it I guess but deleting entire line will be better or am I missing something here?

**Comment by jsumners** [comment]
First thing we should probably do is issue a new v3 minor that adds a depreciation warning.

**Comment by genzyy** [comment]
> First thing we should probably do is issue a new v3 minor that adds a depreciation warning.

For this should the docs be updated [here](https://github.com/fastify/fastify/blob/main/docs/Middleware.md)?

**Comment by jsumners** [comment]
That would be part of it. The main thing would be adding a new deprecation error and triggering it when the method is invoked.

**Comment by mcollina** [comment]
it's already throwing an error, we can just remove it.

**Comment by genzyy** [comment]
> it's already throwing an error, we can just remove it.

Should I make this change on a new branch while keeping `next` as a base and make a pull request to `next` branch?

**Comment by climba03003** [comment]
> > it's already throwing an error, we can just remove it.
> 
> Should I make this change on a new branch while keeping `next` as a base and make a pull request to `next` branch?

Yes

**Comment by genzyy** [comment]
Hey, so to fix this I made a new branch while keeping `next` as the base. But when I run `npm run test`, it gives me some errors. How should I commit the changes?

**Comment by climba03003** [comment]
> 

You need to remove the corresponding test for `.use`.

https://github.com/fastify/fastify/blob/0ae362b7d81b0e823fa7c9c87fc6d4655569d822/test/middleware.test.js#L10-L20
https://github.com/fastify/fastify/blob/0ae362b7d81b0e823fa7c9c87fc6d4655569d822/test/middleware.test.js#L50-L58

**Comment by genzyy** [comment]
> > 
> 
> You need to remove the corresponding test for `.use`.
> 
> https://github.com/fastify/fastify/blob/0ae362b7d81b0e823fa7c9c87fc6d4655569d822/test/middleware.test.js#L10-L20
> 
> https://github.com/fastify/fastify/blob/0ae362b7d81b0e823fa7c9c87fc6d4655569d822/test/middleware.test.js#L50-L58

Removed these tests but I still have errors from multiple test files
![image](https://user-images.githubusercontent.com/48408336/144013746-2a911757-2d2b-4e3d-afe1-fb51530cbe10.png)

**Comment by mcollina** [comment]
> How should I commit the changes?

`git commit -n` to skip the pre-commit hooks.

---

Looking at your errors, I would recommend to `rm -rf node_modules; npm i` as it seems you are missing some deps (joi).

As for the rest, you'd need to look why they are failining.

**Comment by genzyy** [comment]
> > How should I commit the changes?
> 
> `git commit -n` to skip the pre-commit hooks.
> 
> Looking at your errors, I would recommend to `rm -rf node_modules; npm i` as it seems you are missing some deps (joi).
> 
> As for the rest, you'd need to look why they are failining.

Okay, I got here somehow
![image](https://user-images.githubusercontent.com/48408336/144029849-411906d8-db37-4ce2-88aa-136874dc6709.png)

What  about this one?
I removed the two test function mentioned above but I guess more needs to be removed.

**Comment by climba03003** [comment]
> 

Update the `t.plan` to match the actual test number.

**Comment by genzyy** [comment]
> > 
> 
> Update the `t.plan` to match the actual test number.

Fixed all of the above 
But I got these errors popped up

```bash
> fastify@4.0.0-dev test:typescript
> tsc test/types/import.ts && tsd

types/logger.d.ts:18:33 - error TS2694: Namespace 'pino' has no exported member 'SerializerFn'.

18 export type SerializerFn = pino.SerializerFn
                                   ~~~~~~~~~~~~

types/logger.d.ts:23:38 - error TS2694: Namespace 'pino' has no exported member 'BaseLogger'.

23 export type FastifyBaseLogger = pino.BaseLogger & {
                                        ~~~~~~~~~~

types/logger.d.ts:27:34 - error TS2694: Namespace 'pino' has no exported member 'PrettyOptions'.

27 export type PrettyOptions = pino.PrettyOptions & { suppressFlushSyncWarning?: boolean }
                                    ~~~~~~~~~~~~~


Found 3 errors.

```

**Comment by climba03003** [comment]
> > > 
> > 
> > 
> > Update the `t.plan` to match the actual test number.
> 
> Fixed all of the above But I got these errors popped up
> 
> ```shell
> > fastify@4.0.0-dev test:typescript
> > tsc test/types/import.ts && tsd
> 
> types/logger.d.ts:18:33 - error TS2694: Namespace 'pino' has no exported member 'SerializerFn'.
> 
> 18 export type SerializerFn = pino.SerializerFn
>                                    ~~~~~~~~~~~~
> 
> types/logger.d.ts:23:38 - error TS2694: Namespace 'pino' has no exported member 'BaseLogger'.
> 
> 23 export type FastifyBaseLogger = pino.BaseLogger & {
>                                         ~~~~~~~~~~
> 
> types/logger.d.ts:27:34 - error TS2694: Namespace 'pino' has no exported member 'PrettyOptions'.
> 
> 27 export type PrettyOptions = pino.PrettyOptions & { suppressFlushSyncWarning?: boolean }
>                                     ~~~~~~~~~~~~~
> 
> 
> Found 3 errors.
> ```

It is expected to be fail as https://github.com/pinojs/pino/pull/1195 introduce the type breaking in certain area.
You can submit the PR, it should be fixed in another PR.

**Comment by genzyy** [comment]
> It is expected to be fail as [pinojs/pino#1195](https://github.com/pinojs/pino/pull/1195) introduce the type breaking in certain area. You can submit the PR, it should be fixed in another PR.

Got it!
Also we need to add a note in the current version stating the deprecation of `app.use()` in the next release right?

**Comment by jsumners** [comment]
> I think they are useful plugins, we should keep having them but reduce their visibility.

Issues like https://github.com/fastify/fastify-express/issues/61 are why I think we should deprecate the `middie` and `fastify-express` modules. People are not using them to "migrate" to Fastify. They're keeping their old Express stuff around indefinitely because we continue to support it. This is going to continue to cause problems and maintenance burden as these users continue to try and add more and more new technologies on top.

In short: I think we have given enough lifecycles to facilitate migrating. It's time to move forward.

**Comment by Eomm** [comment]
Done in https://github.com/fastify/fastify/pull/3506
